minidump-processor/src/stackwalker/mips.rs

```diff
@@ -0,0 +1,388 @@
+}
+
+fn callee_forwarded_regs(valid: &MinidumpContextValidity) -> HashSet<&'static str> {
+    match valid {
+        MinidumpContextValidity::All => CALLEE_SAVED_REGS.iter().copied().collect(),
+        MinidumpContextValidity::Some(ref which) => CALLEE_SAVED_REGS
+            .iter()
+            .filter(|&reg| which.contains(reg))
+            .copied()
+            .collect(),
+    }
+}
+
+async fn get_caller_by_scan32<P>(
+    ctx: &Mips32Context,
+    callee: &StackFrame,
+    stack_memory: &MinidumpMemory<'_>,
+    modules: &MinidumpModuleList,
+    symbol_provider: &P,
+) -> Option<StackFrame>
+where
+    P: SymbolProvider + Sync,
+{
+    const MAX_STACK_SIZE: u32 = 1024;
+    const MIN_ARGS: u32 = 4;
+    const POINTER_WIDTH: u32 = 4;
+    trace!("trying scan");
+    // Stack scanning is just walking from the end of the frame until we encounter
+    // a value on the stack that looks like a pointer into some code (it's an address
+    // in a range covered by one of our modules). If we find such an instruction,
+    // we assume it's a `ra` value that was saved on the stack by the callee in
+    // its function prologue, following a `jal` (call) instruction of the caller.
+    // The next frame is then assumed to end just before that `ra` value.
+    let valid = &callee.context.valid;
+    let mut last_sp = ctx.get_register(STACK_POINTER, valid)?;
+
+    let mut count = MAX_STACK_SIZE / POINTER_WIDTH;
+    // In case of mips32 ABI the stack frame of a non-leaf function
+    // must have a minimum stack frame size for 4 arguments (4 words).
+    // Move stack pointer for 4 words to avoid reporting non-existing frames
+    // for all frames except the topmost one.
+    // There is no way of knowing if topmost frame belongs to a leaf or
+    // a non-leaf function.
+    if callee.trust != FrameTrust::Context {
+        last_sp += MIN_ARGS * POINTER_WIDTH;
+        count -= MIN_ARGS;
+    }
+
+    for i in 0..count {
+        let address_of_pc = last_sp.checked_add(i * POINTER_WIDTH)?;
+        let caller_pc: u32 = stack_memory.get_memory_at_address(address_of_pc as u64)?;
+        //trace!("unwind: trying addr 0x{address_of_pc:08x}: 0x{caller_pc:08x}");
+        if instruction_seems_valid(caller_pc as u64, modules, symbol_provider).await {
+            // `ra` is usually saved directly at the bottom of the frame,
+            // so sp is just address_of_pc + ptr
+            let caller_sp = address_of_pc.checked_add(POINTER_WIDTH)?;
+
+            // Don't do any more validation, and don't try to restore fp
+            // (that's what breakpad does!)
+
+            trace!(
+                "scan seems valid -- caller_pc: 0x{caller_pc:016x}, caller_sp: 0x{caller_sp:016x}"
+            );
+
+            let mut caller_ctx = MipsContext::default();
+            caller_ctx.set_register(
+                PROGRAM_COUNTER,
+                caller_pc as u64 - (2 * POINTER_WIDTH as u64),
```

Comment:
Wait why are you apply an adjustment to pc here? that's usually the job of "final cleanup" in the outer frame. we store the return-address pc in the context, and `frame.instruction` gets the "adjusted" address we think is the CALL instruction.